bench_ext_test.go

```diff
@@ -0,0 +1,363 @@
+	return Optional2BenchExt{}
+}
+
+func (o Optional2BenchExt) MarshalMsgpack() ([]byte, error) {
+	if o.set {
+		return o.value.MarshalMsgpack()
+	}
+	return NilBytes, nil
+}
+
+func (o *Optional2BenchExt) UnmarshalMsgpack(b []byte) error {
+	if b[0] == msgpcode.Nil {
+		o.set = false
+		return nil
+	}
+	o.set = true
+	return o.value.UnmarshalMsgpack(b)
+}
+
+func (o Optional2BenchExt) EncodeMsgpack(enc *msgpack.Encoder) error {
+	switch {
+	case !o.set:
+		return enc.EncodeNil()
+	default:
+		return enc.Encode(&o.value)
+	}
+}
+
+func (o *Optional2BenchExt) DecodeMsgpack(dec *msgpack.Decoder) error {
+	code, err := dec.PeekCode()
+	if err != nil {
+		return err
+	}
+
+	switch {
+	case code == msgpcode.Nil:
+		o.set = false
+		return nil
+	case msgpcode.IsExt(code):
+		return dec.Decode(&o.value)
+	default:
+		return errors.New("unexpected code")
+	}
+}
+
+type MsgpackExtInterface interface {
+	ExtType() int8
+	msgpack.Marshaler
+	msgpack.Unmarshaler
+}
+
+type OptionalGenericStructWithInterface[T MsgpackExtInterface] struct {
+	value T
+	set   bool
+}
+
+func NewOptionalGenericStructWithInterface[T MsgpackExtInterface](value T) *OptionalGenericStructWithInterface[T] {
+	return &OptionalGenericStructWithInterface[T]{
+		value: value,
+		set:   true,
+	}
+}
+
+func NewEmptyOptionalGenericStructWithInterface[T MsgpackExtInterface]() *OptionalGenericStructWithInterface[T] {
+	return &OptionalGenericStructWithInterface[T]{
+		set: false,
+	}
+}
+
+func (o *OptionalGenericStructWithInterface[T]) DecodeMsgpack(d *msgpack.Decoder) error {
+	code, err := d.PeekCode()
+	if err != nil {
+		return err
+	}
+
+	switch {
+	case code == msgpcode.Nil:
+		o.set = false
+	case msgpcode.IsExt(code):
+		o.set = true
+
+		extID, extLen, err := d.DecodeExtHeader()
+		switch {
+		case err != nil:
+			return err
+		case extID != o.value.ExtType():
+			return errors.New("unexpected extension type")
+		default:
+			ext := make([]byte, extLen)
+
+			err := d.ReadFull(ext)
+			if err != nil {
+				return err
+			}
+
+			err = o.value.UnmarshalMsgpack(ext)
+			if err != nil {
+				return err
+			}
+		}
+
+	default:
+		return errors.New("unexpected type")
+	}
+
+	return nil
+}
+
+func (o *OptionalGenericStructWithInterface[T]) EncodeMsgpack(e *msgpack.Encoder) error {
+	switch {
+	case !o.set:
+		return e.EncodeNil()
+	default:
+		mpdata, err := o.value.MarshalMsgpack()
+		if err != nil {
+			return err
+		}
+
+		err = e.EncodeExtHeader(o.value.ExtType(), len(mpdata))
+		if err != nil {
+			return err
+		}
+
+		mpdataLen, err := e.Writer().Write(mpdata)
+		switch {
+		case err != nil:
+			return err
+		case mpdataLen != len(mpdata):
+			return errors.New("failed to write mpdata")
+		}
+
+		return nil
+	}
+
+}
+
+func BenchmarkExtension(b *testing.B) {
+	msgpack.RegisterExt(8, &BenchExt{})
+
+	var buf bytes.Buffer
+	buf.Grow(4096)
+
+	enc := msgpack.GetEncoder()
+	enc.Reset(&buf)
+
+	dec := msgpack.GetDecoder()
+	dec.Reset(&buf)
+
+	b.Run("Optional1Bench", func(b *testing.B) {
```

Comment:
Does it a go-idiomic practice to name the tests in the manner? Why do we need the CamelCase here? Why not just "optional simple bench" or so on?

Up to @bigbes .

Reply:
Well, I don't have any strong opinion about naming of subtests. On one hand I'm preferring to have name that I can copy to cli to run subtest, but go-idiomatic way is to write clear names with spaces in lowercase.

Let's stick with go-idiomatic way there.